Shopify/shopify-cli
# The `shopify theme push`/`shopify theme serve` commands no longer freeze in some scenarios

### WHY are these changes introduced?

This PR fixes https://github.com/Shopify/shopify-cli/issues/2555, https://github.com/Shopify/shopify-cli/issues/2456, https://github.com/Shopify/cli/issues/570, and some reports we've received on Slack related to the upload process getting frozen at some point.

To handle those reports, we've been recommending the usage of the `--stable` flag. However, this PR aims to avoid that requirement.

This PR moves some components, so it looks a bit larger than it is. Thus, I've split the it into meaningful commits, and I'd recommend the review by following each one of them:

- c7ce5a718744401162eb1fe9d62e96f8d3cb1320 - Introduce `Syncer::Uploader` component
- 8c1afcd1db72d61f7149491d6b85f1d24776a3c8 - Introduce `Syncer::Downloader` component
- fd8c59574e0be36a92e8872e75a1bc2930683594 - Move `/syncer/forms/*` forms to `/syncer/uploader/forms/*`
- 94a9f16aff26ad1b0a08b72c8bdd817e33df098e - Move JSON update/delete handlers from `/syncer/*` to `/syncer/uploader/*`
- b7fb7d2a80350df0929dcbef4ebdb3f042d50f80 - Remove `Theme::ThemeAdminAPIThrottler` component
- 59d26abee77302b373c9c7ab488a8775927af2de - Fix `Theme::IgnoreHelperTest` namespace
- 74257d510bb3977ada5cf42643fd157267925e29 - Introduce `Theme::BackoffHelper` module
- 0c6865869f1b0edbba49379e4245aff1f1a405fa - Introduce `TestHelpers::FakeDB` helper
- 379fbc6e7117a5302b1101f72c0116d976a3d430 - Refactor `Bulk` module to use it as a synchronous module
  - Move `Bulk` from `/theme/theme_admin_api_throttler/*` to `/theme/syncer/uploader/*`
  - Rename `PutRequest` to `BulkItem`
  - Rename `RequestParser` to `BulkRequest`
  - Remove the timeout logic from the `consume_*` method
  - Introduce the timeout logic in the `shutdown` method
  - Introduce the backoff mechanism in the bulk jobs
  - Introduce support to `in_progress_items`
  - Inline `ResponseParser` and `AssetUploadError`
- 26d14a24b0b38294746e0061fdfa6f1448d8ca84 - Update the `Syncer` component to rely on the new `Syncer::Downloader` and `Syncer::Uploader`

### WHAT is this pull request doing?

The core of this change lives in the component responsible for bulk uploads (`Bulk`). We initially tried to incorporate it in the `Syncer` component, so it would make bulk uploads more flexible to other use cases than just uploading. However, the `Syncer` doesn't guarantee the order of uploads, which is critical for creating new themes.

<details data="-----------------------------------------------------------------">
<summary>
<b>Read more</b>
</summary>


This PR extracts sync logic from the `Syncer` as the `Syncer::Uploader` and `Syncer::Downloader` sub-components.

The `Syncer::Uploader` uploads a theme by relying on synchronous steps. That mimics the old behavior of ThemeKit/former-Syncer [1] with the benefits of Bulk uploads.

Finally, this PR updates the `Bulk` module to make it work faster in the new synchronous scenarios. Also, other following enhancements were introduced:
- Support the backoff mechanism in the `Bulk` requests. `Bulk` requests were not reaching the rate limit in a single run. However, with consecutive/parallel executions, the rate limit was impacting bulk requests;
- The new timeout mechanism and the `remaining_items` API [2]. So we avoid the shutdown method holding the execution for too long, and API users know which uploads are still pending.



\-

[1] The former Syncer implementation didn't guarantee the order of uploads. However, as it had a limited number of threads, users never faced any issues with that mechanism.
[2] All requests [at the CLI](https://github.com/Shopify/shopify-cli/blob/main/lib/shopify_cli/http_request.rb#L27-L40) have a timeout of [60 seconds](https://docs.ruby-lang.org/en/master/Net/HTTP.html#attribute-i-read_timeout)

</details>


### How to test your changes?

These changes impact the following commands (click on them to check the test steps):

<details data="-----------------------------------------------------------------">
<summary><b>1. <code>shopify theme push</code></b></summary><br />

Setup a new theme:
```bash
# Create a theme
$ shopify-dev theme init test_theme1

# Go to theme directory:
$ cd test_theme1
```

Create a new theme on your library:
```bash
# Create a new theme on your library
$ shopify-dev theme push -u -t test_theme1

# Create a new theme on your library, but now check the logs
$ DEBUG=1 shopify-dev theme push -u -t test_theme1
```

Create a new theme on your library with the former implementation:
```bash
# Create a new theme on your library with the former implementation
$ shopify-dev theme push -u -t test_theme1 --stable

# Create a new theme on your library with the former implementation, but now check the logs
$ DEBUG=1 shopify-dev theme push -u -t test_theme1 --stable
```

Notice the `--stable` is slower because the logs show it doesn't rely on the bug API
</details>

<details data="-----------------------------------------------------------------">
<summary><b>2. <code>shopify theme serve</code></b></summary><br />

Delete your development theme:
```bash
# Find your development theme ID
$ shopify-dev theme list | grep yours

# Delete it
$ shopify-dev theme delete 1234589
```

Setup a new theme:
```bash
# Create a theme
$ shopify-dev theme init test_theme2

# Go to theme directory:
$ cd test_theme2
```

Run the development server:
```bash
$ DEBUG=1 shopify-dev theme serve
```
Notice that the theme is uploaded using the `Bulk` module and the assets are uploaded using the `Syncer`.

Now, delete your development theme again, and run the development server with the former implementation:
```bash
$ DEBUG=1 shopify-dev theme serve --stable
```
Notice that the theme is uploaded using only the `Syncer`.

Run the development server with the `--theme-editor-sync`:
```bash
$ DEBUG=1 shopify-dev theme serve --theme-editor-sync
```

Notice that the handlers work when:
- The local file is different from the remote file
- The remote file doesn't exist
- The local file doesn't exist

</details>

<details data="-----------------------------------------------------------------">
<summary><b>3. <code>shopify theme pull</code></b></summary><br />

Download a theme from the theme library:
```bash
# Create an empty directory
$ mkdir test_theme3

# Go to theme directory
$ cd test_theme3

# Download a theme from the theme library
$ shopify-dev theme pull -t test_theme1
```
</details>





### Post-release steps

None.

### Update checklist
